library(ggplot2)
library(scales)

# Edit these inputs:


#setwd("~/Desktop/ENTEx/assembly_lengths")

# output prefix
prefix <-"/grid/lippman/data/pansol/stats/Ncharts/Nov30.2022/Contigs"



# All files must be tab-delimited with the chromosome/contig lengths in the second column
# These can be generated by running fastalengths, for instance:
# fastalengths my_reference.fasta > my_reference.lengths


# Reference filename
filename_ref <- "SL4.0.lens" #"GRCh38.lengths"

# Query filenames and names for the plot
query_filenames <-c("Eggplant_V3_Chromosomes.lens",  "Scle2.lens", "Smur2hap2.lens",
"M82.lens", "Setu1.lens", "Spri1.lens",
"Sgig1.lens","Squi2.lens",
"Sabu2.lens","Shav1.lens","Srob1.lens",
"Saet3.lens","Slin1.lens","Stor1.lens",
"Same3.lens","Smac3.lens",  "Scit1.lens", "Spse1.lens", 
"Scan1.lens","Smur2hap1.lens",  "Sang7.lens",
"Solmam1.lens","Solstr1.lens","Solins1.lens","Solang8.lens")
		    #"Sabu2.lens", "Smur2hap1.lens", "Smur2hap2.lens", "Sweet100.lens", "M82.lens") 
#c("Sabu2.lens", "Smur2hap1.lens", "Smur2hap2.lens", "M82.lens", "Sweet100.lens") #c("Squi2.lens", "Sabu2.lens", "Saet3.lens",  "Smac3.lens", "Smur2hap1.lens",  "Smur2hap2.lens",  "Spri1.lens", "Sweet100.lens") #c("falcon_contigs.lengths","megahit_contigs.lengths", "supernova_scaffolds_1.lengths", "supernova_contigs_1.lengths")
query_names <- c("Smel_V3",  "Solcle2","Solmur2hap2",
"Slyc_M82", "Soletu1", "Solpri1",
"Solgig1","Solqui2",
"Solabu2","Solhav1", "Solrob1",
"Solaet3","Sollin1", "Soltor1",
"Solame3","Solmac3",  "Solcit1", "Solpse1",
"Solcan1","Solmur2hap1",  "Solvio1",
"Solmam1","Solstr1","Solins1","Solang8")
		 #"Sabu2", "Smur2hap1", "Smur2hap2", "Sweet100", "SlycM82") 
#c("Squi2", "Sabu2", "Saet3", "Smac3", "Smur2hap1", "Smur2hap2", "Spri1", "Sweet100") #c("FALCON", "Megahit contigs", "Supernova scaffolds", "Supernova contigs")


# first color is the reference, the others are the queries in order
colors <- c("blue", "steelblue", "powderblue" , "mediumaquamarine", "turquoise1",
            "green", "chartreuse4", "chartreuse", "palegreen4",
            "gold", "khaki1", "darkgoldenrod1",
            "coral", "sienna1", "orangered1", "red", "orangered4",
            "violetred", "pink", "plum2", "violet", "purple",
            "brown", "seashell4", "grey", "black" 
	    )
	    #"blue", "purple", "green", "coral", "orange", "mediumaquamarine", "chartreuse", "steelblue", "violet", "red" ) #, "pink", "yellow", "brown", "red") #c("blue","purple","green","coral","orange")

#############################################################################

ref.data <- read.csv(filename_ref, sep="\t", quote='',header=FALSE)
names(ref.data) <- c("name","length")
ref.data$length <- sort(as.numeric(ref.data$length),decreasing=TRUE)

genome.length <- max(sum(ref.data$length))


ref.cumsum <- data.frame(NG=cumsum(ref.data$length) ,contig.length=ref.data$length,contig.source="Reference (SL4.0)")
			 #cumsum(ref.data$length/genome.length*100),contig.length=ref.data$length,contig.source="Reference")

ref.cumsum.0 <- rbind(data.frame(NG=c(0),contig.length=max(ref.cumsum$contig.length),contig.source="Reference (SL4.0)"),ref.cumsum)


for (i in seq(length(query_filenames))) {
  filename_query = query_filenames[i]
  name_query = query_names[i]
  print(paste(filename_query,name_query))

  query.data <- read.csv(filename_query, sep="\t", quote='',header=FALSE)

  names(query.data) <- c("name","length")

  query.data$length <- sort(as.numeric(query.data$length),decreasing=TRUE)

  query.cumsum <- data.frame(NG=cumsum(query.data$length),contig.length=query.data$length,contig.source=name_query)
			     #cumsum(query.data$length/genome.length*100),contig.length=query.data$length,contig.source=name_query)

  query.cumsum.0 <- rbind(data.frame(NG=c(0),contig.length=max(query.cumsum$contig.length),contig.source=name_query),query.cumsum)

  ref.cumsum <- rbind(ref.cumsum,query.cumsum)
  ref.cumsum.0 <- rbind(ref.cumsum.0,query.cumsum.0)
}

bp_format<-function(num) {
  if (num > 1000000000) {
    paste(formatC(num/1000000000,format="f",digits=1,big.mark=",",drop0trailing = TRUE)," Gbp",sep="")
  }
  else if (num > 1000000) {
    paste(formatC(num/1000000,format="f",digits=1,big.mark=",",drop0trailing = TRUE)," Mbp",sep="")
  }
  else {
    paste(formatC(num,format="f",big.mark=",",drop0trailing = TRUE), " bp", sep="")
  }
}

theme_set(theme_bw(base_size = 12) + theme(panel.grid.minor = element_line(colour = NA)))


percent_format <- function(num) {
  paste(num,"%", sep="")
}

png(file=paste(prefix,".Assemblytics.Nchart.png",sep=""),width=1600,height=1000,res=200)

print(
  ggplot(ref.cumsum.0, aes(x = NG, y = contig.length, color=contig.source)) +
    #             ylim(10000000, 1000000000) +
    scale_y_log10(breaks = trans_breaks("log10", function(x) 10^x), labels = trans_format("log10", math_format(10^.x)), limits=c(1000000,genome.length)) +
    scale_x_continuous() +
    geom_path(size=1.5,alpha=0.5) +
    geom_point(data=ref.cumsum, size=2,alpha=0.5) +
    labs(x = paste("Cumulative length (Gb)", sep=""),y="Sequence length",colour="Assembly",title="Cumulative sequence length") +
    scale_color_manual(values=colors) +
    annotation_logticks(sides="lr")
)

dev.off()

